src/snek5000/params.py

```diff
@@ -220,5 +229,275 @@ def _autodoc_par(self, indent=0):
+            )
+
+        # check that no user parameters are overwritten
+        modified_labels = []
+        for index in user_params:
+            try:
+                modified_labels.append(recorded_user_params[index])
+            except KeyError:
+                pass
+        values = user_params.values()
+        for label in modified_labels:
+            if label not in values:
+                raise ValueError(
+                    f"The value {label} would be removed from the user params."
+                )
+
+        reverted = {value: key for key, value in recorded_user_params.items()}
+        for label in user_params.values():
+            try:
+                key = reverted[label]
+            except KeyError:
+                raise ValueError(
+                    f"User parameter {label = } is not already recorded. "
+                    "Use `_record_nek_user_params`"
+                )
+            del recorded_user_params[key]
+
+        recorded_user_params.update(user_params)
+
+    def _save_as_xml(self, path_file=None, comment=None, find_new_name=False):
+        """Invoke :func:`_save_recorded_user_params` and then save to an XML file at ``path_file``."""
+        try:
+            user_params = self.nek.general._recorded_user_params
+        except AttributeError:
+            pass
+        else:
+            if path_file is None:
+                path_dir = Path.cwd()
+            else:
+                path_dir = Path(path_file).parent
+            _save_recorded_user_params(user_params, path_dir)
+
+        return super()._save_as_xml(
+            path_file=path_file, comment=comment, find_new_name=find_new_name
+        )
+
+
+def _save_recorded_user_params(user_params, path_dir):
+    """Save a JSON file from a dictionary denoting ``user_params``"""
+    with open(path_dir / filename_map_user_params, "w") as file:
+        json.dump(user_params, file)
+
+
+def _load_recorded_user_params(path):
+    """Load a JSON file and return a dictionary denoting ``user_params``"""
+    with open(path) as file:
+        tmp = json.load(file)
+    return {int(key): value for key, value in tmp.items()}
+
+
+def _check_and_get_params_nek(params, path):
+    """Check if params is the top level object (via the ``_tag`` attribute) and
+    returen the ``params.nek`` object.
+
+    Parameters
+    ----------
+    params: :class:`Parameters`
+        The ``params`` object
+
+    Returns
+    -------
+    params.nek: :class:`Parameters`
+        The ``params.nek`` object
+
+    """
+    # FIXME: ashwinvis - The path argument is unused. Why do we need it?
+    if not isinstance(path, Path):
+        raise TypeError
+
+    if not isinstance(params, Parameters):
+        raise TypeError
+    if params._tag != "params":
+        raise ValueError(f'{params._tag = } != "params"')
+
+    if params.nek._tag != "nek":
+        raise RuntimeError(f'{params.nek._tag =} != "nek"')
+    return params.nek
```

Comment:
@paugier and here

Reply:
I understood the reason. I will split this function into `_check_path_like` and `_get_params_nek`.

Reply:
Yes, this is what has to be done. It ended up like that but it's unclear.